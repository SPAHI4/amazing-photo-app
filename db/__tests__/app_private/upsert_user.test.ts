import { PoolClient } from 'pg';
import { snapshotSafe, withRootDb } from '../helpers.js';

// This simulates the data your Google API might send
const mockGoogleUser = {
  google_id: 'testGoogleId',
  google_email: 'testEmail@gmail.com',
  google_refresh_token: 'testRefreshToken',
  google_picture_url: 'https://testPictureUrl',
  google_name: 'Test Google User',
  google_given_name: 'Test',
  google_family_name: 'User',
  google_locale: 'en-US',
  google_verified_email: true,
};

async function upsertUser(client: PoolClient, googleUser: any) {
  const {
    rows: [row],
  } = await client.query(
    'SELECT * FROM app_private.upsert_user($1, $2, $3, $4, $5, $6, $7, $8, $9)',
    Object.values(googleUser),
  );

  return row;
}

describe('Upsert User', () => {
  it('upserts a new user correctly', () =>
    withRootDb(async (client) => {
      const user = await upsertUser(client, mockGoogleUser);
      expect(user).toBeTruthy();
      expect(user.id).toBeTruthy();
      expect(snapshotSafe(user)).toMatchSnapshot();
    }));

  it('updates an existing user correctly', () =>
    withRootDb(async (client) => {
      await upsertUser(client, mockGoogleUser);
      const updatedUser = { ...mockGoogleUser, google_name: 'Updated name' };
      const user = await upsertUser(client, updatedUser);
      expect(user).toBeTruthy();
      expect(user.display_name).toEqual(updatedUser.google_name);
      expect(snapshotSafe(user)).toMatchSnapshot();
    }));
});
