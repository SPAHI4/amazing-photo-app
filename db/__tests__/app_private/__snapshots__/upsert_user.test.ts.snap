// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upsert User updates an existing user correctly 1`] = `
{
  "archived_at": null,
  "created_at": "[DATE]",
  "display_name": "Updated name",
  "id": "[ID]",
  "is_archived": false,
  "picture_url": "https://testPictureUrl",
  "role": "app_user",
  "updated_at": "[DATE]",
}
`;

exports[`Upsert User upserts a new user correctly 1`] = `
{
  "archived_at": null,
  "created_at": "[DATE]",
  "display_name": "Test Google User",
  "id": "[ID]",
  "is_archived": false,
  "picture_url": "https://testPictureUrl",
  "role": "app_user",
  "updated_at": "[DATE]",
}
`;
